File: server/camcops_server/tasks/sfmpq2.py
#!/usr/bin/env python

"""
camcops_server/tasks/sfmpq2.py

===============================================================================

    Copyright (C) 2012, University of Cambridge, Department of Psychiatry.
    Created by Rudolf Cardinal (rnc1001@cam.ac.uk).

    This file is part of CamCOPS.

    CamCOPS is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    CamCOPS is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with CamCOPS. If not, see <https://www.gnu.org/licenses/>.

===============================================================================

**Short-Form McGill Pain Questionnaire (SF-MPQ2) task.**

"""

from camcops_server.cc_modules.cc_constants import CssClass
from camcops_server.cc_modules.cc_html import tr_qa, tr, answer
from camcops_server.cc_modules.cc_request import CamcopsRequest
from camcops_server.cc_modules.cc_sqla_coltypes import (
    CamcopsColumn,
    ZERO_TO_10_CHECKER,
)

from camcops_server.cc_modules.cc_summaryelement import SummaryElement
from camcops_server.cc_modules.cc_task import TaskHasPatientMixin, Task
import cardinal_pythonlib.rnc_web as ws
from cardinal_pythonlib.stringfunc import strseq
from sqlalchemy import Float, Integer
from sqlalchemy.ext.declarative import DeclarativeMeta
from typing import List, Type, Tuple, Dict, Any


class Sfmpq2Metaclass(DeclarativeMeta):
    # noinspection PyInitNewSignature
    def __init__(
        cls: Type["Sfmpq2"],
        name: str,
        bases: Tuple[Type, ...],
        classdict: Dict[str, Any],
    ) -> None:

        # Field descriptions are open access, as per:
        # https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5221718/
        # https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3225325/
        comment_strings = [
            "throbbing",
            "shooting",
            "stabbing",
            "sharp",
            "cramping",
            "gnawing",
            "hot-burning",
            "aching",
            "heavy",
            "tender",
            "splitting",
            "tiring–exhausting",
            "sickening",
            "fearful",
            "punishing–cruel",
            "electric-shock",
            "cold-freezing",
            "piercing",
            "light touch",
            "itching",
            "tingling",
            "numbness",
        ]
        score_comment = "(0 none - 10 worst)"

        for q_index in range(0, cls.N_QUESTIONS):
            q_num = q_index + 1
            q_field = "q{}".format(q_num)

            setattr(
                cls,
                q_field,
                CamcopsColumn(
                    q_field,
                    Integer,
                    permitted_value_checker=ZERO_TO_10_CHECKER,
                    comment="Q{} ({}) {}".format(
                        q_num, comment_strings[q_index], score_comment
                    ),
                ),
            )

        super().__init__(name, bases, classdict)


class Sfmpq2(TaskHasPatientMixin, Task, metaclass=Sfmpq2Metaclass):
    __tablename__ = "sfmpq2"
    shortname = "SF-MPQ2"

    N_QUESTIONS = 22
    MAX_SCORE_PER_Q = 10
    ALL_QUESTIONS = strseq("q", 1, N_QUESTIONS)

    CONTINUOUS_PAIN_QUESTIONS = Task.fieldnames_from_list(
        "q", {1, 5, 6, 8, 9, 10}
    )
    INTERMITTENT_PAIN_QUESTIONS = Task.fieldnames_from_list(
        "q", {2, 3, 4, 11, 16, 18}
    )
    NEUROPATHIC_PAIN_QUESTIONS = Task.fieldnames_from_list(
        "q", {7, 17, 19, 20, 21, 22}
    )
    AFFECTIVE_PAIN_QUESTIONS = Task.fieldnames_from_list("q", {12, 13, 14, 15})

    @staticmethod
    def longname(req: CamcopsRequest) -> str:
        _ = req.gettext
        return _("Short-Form McGill Pain Questionnaire 2")

    def get_summaries(self, req: CamcopsRequest) -> List[SummaryElement]:
        return self.standard_task_summary_fields() + [
            SummaryElement(
                name="total_pain",
                coltype=Float(),
                value=self.total_pain(),
                comment=f"Total pain (/{self.MAX_SCORE_PER_Q})",
            ),
            SummaryElement(
                name="continuous_pain",
                coltype=Float(),
                value=self.continuous_pain(),
                comment=f"Continuous pain (/{self.MAX_SCORE_PER_Q})",
            ),
            SummaryElement(
                name="intermittent_pain",
                coltype=Float(),
                value=self.intermittent_pain(),
                comment=f"Intermittent pain (/{self.MAX_SCORE_PER_Q})",
            ),
            SummaryElement(
                name="neuropathic_pain",
                coltype=Float(),
                value=self.neuropathic_pain(),
                comment=f"Neuropathic pain (/{self.MAX_SCORE_PER_Q})",
            ),
            SummaryElement(
                name="affective_pain",
                coltype=Float(),
                value=self.affective_pain(),
                comment=f"Affective pain (/{self.MAX_SCORE_PER_Q})",
            ),
        ]

    def is_complete(self) -> bool:
        if self.any_fields_none(self.ALL_QUESTIONS):
            return False
        if not self.field_contents_valid():
            return False
        return True

    def total_pain(self) -> float:
        return self.mean_fields(self.ALL_QUESTIONS)

    def continuous_pain(self) -> float:
        return self.mean_fields(self.CONTINUOUS_PAIN_QUESTIONS)

    def intermittent_pain(self) -> float:
        return self.mean_fields(self.INTERMITTENT_PAIN_QUESTIONS)

    def neuropathic_pain(self) -> float:
        return self.mean_fields(self.NEUROPATHIC_PAIN_QUESTIONS)

    def affective_pain(self) -> float:
        return self.mean_fields(self.AFFECTIVE_PAIN_QUESTIONS)

    def format_average(self, value) -> str:
        return "{} / {}".format(
            answer(ws.number_to_dp(value, 3, default="?")),
            self.MAX_SCORE_PER_Q,
        )

    def get_task_html(self, req: CamcopsRequest) -> str:
        rows = ""
        for q_num in range(1, self.N_QUESTIONS + 1):
            q_field = "q" + str(q_num)
            question_cell = "{}. {}".format(q_num, self.wxstring(req, q_field))

            score = getattr(self, q_field)

            rows += tr_qa(question_cell, score)

        html = """
            <div class="{CssClass.SUMMARY}">
                <table class="{CssClass.SUMMARY}">
                    {tr_is_complete}
                    {total_pain}
                    {continuous_pain}
                    {intermittent_pain}
                    {neuropathic_pain}
                    {affective_pain}
                </table>
            </div>
            <table class="{CssClass.TASKDETAIL}">
                <tr>
                    <th width="60%">Question</th>
                    <th width="40%">Answer <sup>[6]</sup></th>
                </tr>
                {rows}
            </table>
            <div class="{CssClass.FOOTNOTES}">
                [1] Average of items 1–22.
                [2] Average of items 1, 5, 6, 8, 9, 10.
                [3] Average of items 2, 3, 4, 11, 16, 18.
                [4] Average of items 7, 17, 19, 20, 21, 22.
                [5] Average of items 12, 13, 14, 15.
                [6] All items are rated from “0 – none” to
                    “10 – worst possible”.
            </div>
        """.format(
            CssClass=CssClass,
            tr_is_complete=self.get_is_complete_tr(req),
            total_pain=tr(
                self.wxstring(req, "total_pain") + " <sup>[1]</sup>",
                self.format_average(self.total_pain()),
            ),
            continuous_pain=tr(
                self.wxstring(req, "continuous_pain") + " <sup>[2]</sup>",
                self.format_average(self.continuous_pain()),
            ),
            intermittent_pain=tr(
                self.wxstring(req, "intermittent_pain") + " <sup>[3]</sup>",
                self.format_average(self.intermittent_pain()),
            ),
            neuropathic_pain=tr(
                self.wxstring(req, "neuropathic_pain") + " <sup>[4]</sup>",
                self.format_average(self.neuropathic_pain()),
            ),
            affective_pain=tr(
                self.wxstring(req, "affective_pain") + " <sup>[5]</sup>",
                self.format_average(self.affective_pain()),
            ),
            rows=rows,
        )
        return html

File: tablet_qt/tasks/sfmpq2.cpp
/*
    Copyright (C) 2012, University of Cambridge, Department of Psychiatry.
    Created by Rudolf Cardinal (rnc1001@cam.ac.uk).

    This file is part of CamCOPS.

    CamCOPS is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    CamCOPS is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with CamCOPS. If not, see <https://www.gnu.org/licenses/>.
*/

#include "sfmpq2.h"

#include "lib/stringfunc.h"
#include "lib/uifunc.h"
#include "maths/mathfunc.h"
#include "questionnairelib/namevalueoptions.h"
#include "questionnairelib/namevaluepair.h"
#include "questionnairelib/questionnaire.h"
#include "questionnairelib/qumcqgrid.h"
#include "questionnairelib/qutext.h"
#include "tasklib/taskfactory.h"
#include "tasklib/taskregistrar.h"
using mathfunc::anyNull;
using mathfunc::meanOrNull;
using mathfunc::scorePhraseVariant;
using stringfunc::strnum;
using stringfunc::strnumlist;
using stringfunc::strseq;

const int FIRST_Q = 1;
const int N_QUESTIONS = 22;
const int MAX_SCORE_PER_Q = 10;
const QString QPREFIX("q");

const QVector<int> CONTINUOUS_PAIN_QUESTIONS{1, 5, 6, 8, 9, 10};
const QVector<int> INTERMITTENT_PAIN_QUESTIONS{2, 3, 4, 11, 16, 18};
const QVector<int> NEUROPATHIC_PAIN_QUESTIONS{7, 17, 19, 20, 21, 22};
const QVector<int> AFFECTIVE_PAIN_QUESTIONS{12, 13, 14, 15};

const bool IGNORE_NULL_FOR_MEAN = true;


const QString Sfmpq2::SFMPQ2_TABLENAME("sfmpq2");

void initializeSfmpq2(TaskFactory& factory)
{
    static TaskRegistrar<Sfmpq2> registered(factory);
}

Sfmpq2::Sfmpq2(CamcopsApp& app, DatabaseManager& db, const int load_pk) :
    Task(app, db, SFMPQ2_TABLENAME, false, false, false),
    // ... anon, clin, resp
    m_questionnaire(nullptr)
{
    addFields(
        strseq(QPREFIX, FIRST_Q, N_QUESTIONS), QMetaType::fromType<int>()
    );

    load(load_pk);  // MUST ALWAYS CALL from derived Task constructor.
}

// ============================================================================
// Class info
// ============================================================================

QString Sfmpq2::shortname() const
{
    return "SF-MPQ2";
}

QString Sfmpq2::longname() const
{
    return tr("Short-Form McGill Pain Questionnaire 2");
}

QString Sfmpq2::description() const
{
    return tr(
        "22-item self-report measure of pain symptoms of both "
        "neuropathic and non-neuropathic pain conditions."
    );
}

QStringList Sfmpq2::fieldNames() const
{
    return strseq(QPREFIX, FIRST_Q, N_QUESTIONS);
}

// ============================================================================
// Instance info
// ============================================================================

bool Sfmpq2::isComplete() const
{
    if (anyNull(values(fieldNames()))) {
        return false;
    }

    return true;
}

QVariant Sfmpq2::totalPain() const
{
    return meanOrNull(values(fieldNames()), IGNORE_NULL_FOR_MEAN);
}

QVariant Sfmpq2::continuousPain() const
{
    return meanOrNull(
        values(strnumlist(QPREFIX, CONTINUOUS_PAIN_QUESTIONS)),
        IGNORE_NULL_FOR_MEAN
    );
}

QVariant Sfmpq2::intermittentPain() const
{
    return meanOrNull(
        values(strnumlist(QPREFIX, INTERMITTENT_PAIN_QUESTIONS)),
        IGNORE_NULL_FOR_MEAN
    );
}

QVariant Sfmpq2::neuropathicPain() const
{
    return meanOrNull(
        values(strnumlist(QPREFIX, NEUROPATHIC_PAIN_QUESTIONS)),
        IGNORE_NULL_FOR_MEAN
    );
}

QVariant Sfmpq2::affectivePain() const
{
    return meanOrNull(
        values(strnumlist(QPREFIX, AFFECTIVE_PAIN_QUESTIONS)),
        IGNORE_NULL_FOR_MEAN
    );
}

QStringList Sfmpq2::summary() const
{
    return QStringList{
        scorePhraseVariant(
            xstring("total_pain"), totalPain(), MAX_SCORE_PER_Q
        ),
        scorePhraseVariant(
            xstring("continuous_pain"), continuousPain(), MAX_SCORE_PER_Q
        ),
        scorePhraseVariant(
            xstring("intermittent_pain"), intermittentPain(), MAX_SCORE_PER_Q
        ),
        scorePhraseVariant(
            xstring("neuropathic_pain"), neuropathicPain(), MAX_SCORE_PER_Q
        ),
        scorePhraseVariant(
            xstring("affective_pain"), affectivePain(), MAX_SCORE_PER_Q
        ),
    };
}

QStringList Sfmpq2::detail() const
{
    QStringList lines = completenessInfo();
    const QString spacer = " ";
    const QString suffix = "";
    lines
        += fieldSummaries("q", suffix, spacer, QPREFIX, FIRST_Q, N_QUESTIONS);
    lines.append("");
    lines += summary();

    return lines;
}

OpenableWidget* Sfmpq2::editor(const bool read_only)
{
    NameValueOptions intensity_options;
    QVector<int> option_widths;

    for (int i = 0; i <= MAX_SCORE_PER_Q; i++) {
        QString xstringname = QString("a%1").arg(i);
        auto option = NameValuePair(xstring(xstringname), i);
        intensity_options.append(option);
        option_widths.append(3);
    };

    QVector<QuestionWithOneField> q_field_pairs;

    for (const QString& fieldname : fieldNames()) {
        const QString& description = xstring(fieldname);
        q_field_pairs.append(
            QuestionWithOneField(description, fieldRef(fieldname))
        );
    }
    auto grid = new QuMcqGrid(q_field_pairs, intensity_options);

    const int question_width = 4;
    grid->setWidth(question_width, option_widths);

    // Repeat options every five lines
    QVector<McqGridSubtitle> subtitles{
        {5, ""},
        {10, ""},
        {15, ""},
        {20, ""},
    };
    grid->setSubtitles(subtitles);

    QuPagePtr page((new QuPage{new QuText(xstring("instructions")), grid})
                       ->setTitle(xstring("title_main")));

    m_questionnaire = new Questionnaire(m_app, {page});
    m_questionnaire->setType(QuPage::PageType::Patient);
    m_questionnaire->setReadOnly(read_only);

    return m_questionnaire;
}
